---
title: " **Analysis of the level of DNA methylation in a dataset with R**"
author: Federica Brando
date: 20/06/20
abstract: "This report describes the pipeline for module 2 final assignment of DNA and RNA dynamics course of University of Bologna master's degree in Bioinformatics. Each main chapter corresponds to each step, briefly commented and explained. The dataset consists in eight samples: 4 belonging to down syndrome ```DS``` group and four wild-type ```WT```. The aim of this report is to illustrate how to perform the analysis of the level of methylation in a given dataset with R."
output:
    html_notebook:
        css: style.css
        toc: true
        toc_depht: true
        toc_float: true
        number_sections: true
---

# **Loading raw data**
The first step of the workflow is to load and read raw data with ```minfi``` library. ```minfi``` is a very useful and complete library of tools for analyzing and visualizing Illumina methylation array data with a special focus on the 450k array. So, we load the ```Illumina450Manifest_clean.RData``` and read the data stored in ```./Input_data``` directory with ```minfi``` function ```read.metharray.sheet()``` .
```{r step 1 - reading samplesheet, message = FALSE, warning = FALSE}
rm(list=ls())  # clean the workspace
setwd("~/PycharmProjects/drd/Report")  # set the working directory
library(minfi)  # load minfi package
load('~/PycharmProjects/drd/Lesson2/Illumina450Manifest_clean.RData')  # load the illumina manifest

baseDir <- ("~/PycharmProjects/drd/Report/Input_data")  # set the data directory
targets <- read.metharray.sheet(baseDir)  # reading the sample sheet containing pheno-data
targets
```
Then we create a variable in which we store the ```RGChannelSet``` object.
```{r step 1 - RGset, message = FALSE, warning = FALSE}
# reading methylation array data {class:RGChannelSet}
RGset <- read.metharray.exp(targets = targets)
RGset
```
# **Red and Green fluorescence**
The second step is to create a ```data.frame``` for Red and Green fluorescence respectively.
```{r step 2 RG dataframes, message = FALSE, warning = FALSE}
# Red dataframe
Red <- data.frame(getRed(RGset))
head(Red)

# Green dataframe
Green <- data.frame(getGreen(RGset))
head(Green)
```
# **Fluorescence inspection**
>**Q**: What are the Red and Green fluorescence for ```42639338```?

In order to answer this question we can inspect the ```data.frame``` just created with the following:
```{r step 3, message = FALSE, warning = FALSE}
Red[rownames(Red)=="42639338",]

Green[rownames(Green)=="42639338",]
```
## _Optional_: Design probe inspection
We can retrieve information on the probe design by checking the address we previously used to the manifest file and eventually, if design probe is Type I, report the color.
```{r step 3 Design probe inspection, message = FALSE, warning = FALSE}
Addr42639338 <- Illumina450Manifest_clean[Illumina450Manifest_clean$AddressA_ID=="42639338",]
Addr42639338$Infinium_Design_Type
```
The result shows that the probe design is a type II probe, therefore we would not find a color assigned to it.

## Table recap

Sample|Slide Array | Red fl. | Green fl. | Design Type|Color|
|:----|:-----:|:-------:|:---------:|:----------:|:---:|
1020|X5775278051_R01C01| 14046 | 12057 | II| N/A|
1036|X5775278051_R04C02| 17095 | 12635 | II| N/A|
3038|X5775278078_R02C01| 10490 | 10730 | II| N/A|
3042|X5775278078_R05C01| 14706 | 12498 | II| N/A|
3052|X5775278078_R05C02| 12822 | 12067 | II| N/A|
1016|X5930514034_R01C02| 13961 | 10900 | II| N/A|
1029|X5930514035_R04C03| 16314 | 12788 | II| N/A|
3029|X5930514035_R06C02| 15172 | 14474 | II| N/A|

# **Creating methylation signal object**

Next step is to convert the ```RGChannelSet``` object into a ```MethylSet``` object. The latter contains the methylated and unmethylated signals of our data. We use the ```preprocessRaw()``` function to perform this task.
```{r step 4, message = FALSE, warning = FALSE}
MSet.raw <- preprocessRaw(RGset)  # Create a MSet.raw object
MSet.raw
```
# **Quality check**
Next step in the workflow is to evaluate the _quality_ of the data. We can do this by performing three types of quality checks:


1. __QCplot__
2. __Negative controls intensity check__
3. __Detection of pValues__

## _QCplot_
A QCplot is a plot in which we plot the median of the methylation signal and the median of the unmethylation signal for each sample. If the samples have good quality, high medians for both methylation and unmethylation signals, they would be found in the upper part of the plot, otherwise we find them in the lower region of the plot and we consider them as low-quality samples.

We use the ```getQC()``` function to extract the aforementioned median values and ```plotQC()``` function to plot them.
```{r step 5 QCplot, fig.align="center", fig.width=5, fig.height=5, fig.cap="**Figure 1**: QC plot", message = FALSE, warning = FALSE}
qc <- getQC(MSet.raw)
plotQC(qc)
```
In this case we can affirm that samples have **good quality**, indeed they are located in the upper region of the plot, so they have high median methylation and unmethylation values.

## _Negative control intensity check_
These check is useful to understand the quality of negative controls. If it returns high green or red fluorescence then the background fluorescence is indistinguishable from the real signal and indicates poor sample quality. The background level intensity is at 10 ( $\log{2}$ ), therefore below that threshold we consider the samples to have a good quality. otherwise we don't.

We use the function ```controlStripPlot()``` to obtain the result (**Fig.2** below). We can also observe that the samples maintain a consistent intensity variation between slides, that is reproducible for both red and green, and they are all below background level (10).
```{r step 5 Negative control, fig.align="center", fig.width=5, fig.height=5, fig.cap="**Figure 2**: Negative control check.", message = FALSE, warning = FALSE}
controlStripPlot(RGset, controls="NEGATIVE")
```
## _Detection of pValues_
>**Q**: for each sample, how many probes have a detection p-value higher than the threshold ```0.05```?

The detection of p-values points out the probability that the signal we are observing in the data is different from the background. The null hypotheses states that signal are equal, the alternative hypotheses states that the signals differ from each other. Indeed, small p-values indicate good position, we can reject the null hypotheses and state that the signal is different from the background with high confidence.

To perform this task we use ```detectionP()``` function on ```RGset``` to obtain the p-Values. We store the probes with p-Value higher than the assigned threshold in ```failed``` variable and then we inspect it to retrieve the number of those probes per sample.
```{r step 5 pValues detection, message = FALSE, warning = FALSE}
detP <- detectionP(RGset)

failed <- detP>0.05
summary(failed)  # True probes have a p-value > 0.05
```

### Table recap

Sample| Slide Array | Failed (p-value>0.05)|
|:---:|:-----------:|:---------------------|
1020|5775278051_R01C01 | 247 |
1036|5775278051_R04C02 | 210 |
3028|5775278078_R02C01 | 264 |
3042|5775278078_R05C01 | 413 |
3052|5775278078_R05C02 | 385 |
1016|5930514034_R01C02 | 91 |
1029|5930514035_R04C02 | 46 |
3029|5930514035_R06C02 | 115 |

# **Computing raw $\beta$ and M values**

Raw $\beta$  and M are computed as: $\beta = \frac{M}{M+U}$ and $M = \log_{2}(\frac{M}{U})$. $\beta$ is a value that expresses the level of methylation at a given CpG site, it ranges between 0 and 1, where 1 stands for completely methylated and 0 stands for the opposite. M, instead, is a variable that can take any value of the real line.

We can plot the densities of continuous variable of $\beta$ and M to observe the level of methylation in the sample. In our sample we can subset the dataset accordingly to the groups: ```DS``` and ```WT```.

In order to extract $\beta$ and M values we can use the functions ```getBeta()``` and ```getM()``` respectively, on both ```DS``` and ```WT``` subsets. We then compute ```mean``` and ```density``` in order to plot the results.
```{r step 6 - reading data, message = FALSE, warning = FALSE}
# reading the data
csvtargets <- read.csv("Input_data/Samplesheet_report_2020.csv")

#WT
WT <- csvtargets[csvtargets$Group=="WT", "Basename"]
WTSet <- MSet.raw[,colnames(MSet.raw) %in% WT]
BetaWT <- getBeta(WTSet)  # getting beta values
M_WT <- getM(WTSet)  # getting M values
meanBeta_WT <- apply(BetaWT, 1, mean, na.rm=T)
meanM_WT <- apply(M_WT, 1, mean, na.rm=T)
d_meanBeta_WT <- density(meanBeta_WT)
d_meanM_WT <- density(meanM_WT)

#DS
DS <- csvtargets[csvtargets$Group=="DS", "Basename"]
DSSet <- MSet.raw[,colnames(MSet.raw) %in% DS]
BetaDS <- getBeta(DSSet)
M_DS <- getM(DSSet)
meanBeta_DS <- apply(BetaDS, 1, mean, na.rm=T)
meanM_DS <- apply(M_DS, 1, mean, na.rm=T)
d_meanBeta_DS <- density(meanBeta_DS)
d_meanM_DS <- density(meanM_DS)

```
```{r step 6 plot, fig.align="center", fig.cap=" **Figure 3**: (left) beta plot for subset WT and DS, (right) M plot for subset WT and DS", message = FALSE, warning = FALSE}
par(mfrow=c(1,2))
plot(d_meanBeta_WT,main="Density of Beta Values",col="blue", ylim = c(-0.01, 3.9))
lines(d_meanBeta_DS, col="red")
legend("topright", legend=c("WT", "DS"),
       col=c("blue", "red"), lty=1:1, cex=0.5)
plot(d_meanM_WT,main="Density of M Values",col="blue")
lines(d_meanM_DS, col='red')
legend("topright", legend=c("WT", "DS"),
       col=c("blue", "red"), lty=1:1, cex=0.5)
```
The inspection of the plot shows that the two subsets ```WT``` and ```DS``` have similar distribution. We can note a slightly higher peak for ```DS``` group for the beta values, other that that they follow the same trend.

# **Normalization**
Data as it is right now cannot be compared, therefore we need to normalize it in order to get rid of technical variation and systematic bias. We normalize using the function ```preprocessQuantile()```, this function is a stratified quantile normalization, it stratifies the probe by region and returns the $\beta$ values normalized. We, then, compare raw data and normalize data from Infinium I and II design.

First, we retrieve all the raw $\beta$ values from ```MSet.raw```, then we normalize data from ```RGset``` and extract normalized $\beta$ from it.
```{r step 7, message = FALSE, warning = FALSE}
beta <- getBeta(MSet.raw)

preprocessQuantile_results <- preprocessQuantile(RGset)
preprocessQuantileBeta <- getBeta(preprocessQuantile_results)
```
Then we extract Infinium I and II probe **raw** and **normalized** $\beta$.

## _Raw data_
Infinium design is stored under the ```Infinium_Design_Type``` entry in the ```Illumina450Manifest_clean```. We
 extract raw data from there and cross-reference the data from the ```beta``` variable that stores raw $\beta$ values
  by searching for the Illumina ID (```IlmnID```). Then we compute ```mean``` and ```standard deviation``` on both
   subsets. We compute also the density for plot purposes.
```{r step 7 InfI raw, message = FALSE, warning = FALSE}
# Infinium design I
InfI_df <- Illumina450Manifest_clean[Illumina450Manifest_clean$Infinium_Design_Type=="I",]
InfI_df <- droplevels(InfI_df)
# cross-reference between InfI_df and raw beta
BetaI <- beta[rownames(beta) %in% InfI_df$IlmnID,]
# computing mean and sd
meanBeta_I <- apply(BetaI,1,mean)
sdBeta_I <- apply(BetaI, 1, sd)
# computing densities
d_meanBetaI <- density(meanBeta_I, na.rm=T)
d_sdBetaI <- density(sdBeta_I,na.rm=T)
```

We do the same with ```Infinium design II```.
```{r step 7 InfII raw, message = FALSE, warning = FALSE}
# Infinium design II
InfII_df <- Illumina450Manifest_clean[Illumina450Manifest_clean$Infinium_Design_Type=="II",]
InfII_df <- droplevels(InfII_df)
# cross-reference between InfII_df and raw beta
BetaII <- beta[rownames(beta) %in% InfII_df$IlmnID,]
# computing mean and sd
meanBeta_II <- apply(BetaII,1,mean)
sdBeta_II <- apply(BetaII,1,sd)
# computing the densities
d_meanBetaII <- density(meanBeta_II, na.rm=T)
d_sdBetaII <- density(sdBeta_II, na.rm=T)
```
## _Normalized data_
The normalization process consists in the cross-reference of the ```IlmnID``` associated to the Infinium design type
 and the ```preprocessQuantileBeta``` variable that stores normalized $\beta$ values. Then ```mean``` and ```standard
  deviation``` are computed on both subsets. We compute also the density in order to generate a plot later.
```{r step 7 InfI norm, message = FALSE, warning = FALSE}
# Infinium design I
# cross-reference between InfI_df and normalized beta
preprocessQuantileBeta_I <- preprocessQuantileBeta[rownames(preprocessQuantileBeta) %in% InfI_df$IlmnID,]
# computing mean and sd
meanBeta_preprocessQuantile_I <- apply(preprocessQuantileBeta_I, 1, mean)
sdBeta_preprocessQuantile_I <- apply(preprocessQuantileBeta_I, 1, sd)
# computing densities
d_meanBeta_preprocessQuantile_I <- density(meanBeta_preprocessQuantile_I, na.rm=T)
d_sdBeta_preprocessQuantile_I <- density(sdBeta_preprocessQuantile_I, na.rm=T)
```

We do the same for ```Infinium design II```
```{r step 7 InfII norm, message = FALSE, warning = FALSE}
# Infinium design II
# cross-reference between InfII_df and normalized beta
preprocessQuantileBeta_II <- preprocessQuantileBeta[rownames(preprocessQuantileBeta) %in% InfII_df$IlmnID,]
# computing mean and sd
meanBeta_preprocessQuantile_II <- apply(preprocessQuantileBeta_II, 1, mean)
sdBeta_preprocessQuantile_II <- apply(preprocessQuantileBeta_II, 1, sd)
# computing the densities
d_meanBeta_preprocessQuantile_II <- density(meanBeta_preprocessQuantile_II, na.rm=T)
d_sdBeta_preprocessQuantile_II <- density(sdBeta_preprocessQuantile_II, na.rm=T)
```
## _Plotting raw and normalized data_
We plot the results in order to compare Infinium I and Infinium II design raw and normalized $\beta$ for both subsets. Also a ```boxplot``` is plotted in order to observe changing in the distribution of the overall data.
```{r step 7 plot raw, fig.width=8, fig.height=10, fig.cap = "**Figure 4**: Plots on the right represent _raw_ data while plots on the left represent _normalized_ data", message = FALSE, warning = FALSE}
# Mean beta raw
par(mfrow=c(3,2), pin = c(2,2))
plot(d_meanBetaI, col = "green", main = "Mean Beta (raw data)", xlim=c(-0.1,1.1),ylim = c(-0.5,6))
lines(d_meanBetaII, col = "orange")
legend("topright", legend=c("InfI", "InfII"),
       col=c("green", "orange"), lty=1:1, cex=0.9)
# Mean beta norm
plot(d_meanBeta_preprocessQuantile_I,col="blue",main=" Mean Beta (normalized data)", xlim = c(-0.1,1.1), ylim = c(-0.5
  ,6))
lines(d_meanBeta_preprocessQuantile_II,col="red")
legend("topright", legend=c("InfI", "InfII"),
       col=c("blue", "red"), lty=1:1, cex=0.9)
# sd beta raw
plot(d_sdBetaI, col = "green", main = "StDev Beta (raw data)", xlim = c(-0.05,0.45), ylim = c(-5,90))
lines(d_sdBetaII, col = "orange")
legend("topright", legend=c("InfI", "InfII"),
       col=c("green", "orange"), lty=1:1, cex=0.9)
# sd beta norm
plot(d_sdBeta_preprocessQuantile_I,col="blue",main="StDev Beta (normalized beta)", xlim = c(-0.05,0.45), ylim = c(-5
  ,90))
lines(d_sdBeta_preprocessQuantile_II,col="red")
legend("topright", legend=c("InfI", "InfII"),
       col=c("blue", "red"), lty=1:1, cex=0.9)
# beta raw
boxplot(beta,main="raw Beta boxplot")
# beta norm
boxplot(preprocessQuantileBeta, main = "preprocessQuantile Beta boxplot")
```
The most evident change we can observe in the plots (**Fig.4**) is between raw and normalized $\beta$ represented by the boxplot. After normalization we can note how the distributions are behaving almost identically. The only feature that behaves slightly different is the median. The plots referring to the density of the mean differ from each other in several ways: Infinium II raw mean beta distribution is shifted toward the center, while Infinium I raw mean beta distribution is less centered and closer to the edge of the graph. After normalization we can note how Infinium I mean distribution gets corrected and the first peak is reduced. Standard deviation has quite similar behaviour to the previous case: Infinium I raw standard deviation distribution has quite high peak, while Infinium II raw standard deviation distribution has a lower peak but more spread. After the normalization process with ```preprocessQuantile()``` function, we can notice a slightly increasing and shrinking of Infinium II standard deviation distribution behaviour and an evident reduction of the peak for Infinium I design. All those observed difference are due to the different chemistries that operate each design, indeed infinium I probes tend to have a larger span for beta values than Infinium II probe.

#  **PCA**
Next step we perform a PCA on the $\beta$ matrix generated in the normalization step. Principal components analysis
 is a method used with methylation data to identify predominant pattern of methylation along with the identified
  components, in this way samples with similar patterns clusterize together.

  We use ```prcomp()``` function to extract the principal components.

```{r step 8 pca - result, warning = FALSE, message = FALSE }
pca_results <- prcomp(t(preprocessQuantileBeta),scale=T)
print(summary(pca_results))
```
The screeplot (obtained with ```fviz_eig()``` function) shows that the variance is descending through the components in a consistent manner. Indeed the proportion of the components is fairly similar, the first 7 components account for the total of the variance (100%). Plotting the first two components would account only for almost 40% of the total variance, it could be not
   enough to make assumptions.
```{r step 8 screeplot, fig.width=7, fig.height=7, fig.cap = " **Figure 5** : Screeplot of PCA results", warning = FALSE, message = FALSE }
library(factoextra)
fviz_eig(pca_results, addlabels = TRUE, barfill = "grey", barcolor = "black")
```
```{r step 8 pca plot, centered, fig.width=7, fig.height=7, fig.cap = " **Figure 6** : PCA results" }
group <- factor(targets$Group)
palette(c("red", "blue"))
plot(pca_results$x[,1], pca_results$x[,2],cex=2,pch=19, col=c(group),xlab="PC1",ylab="PC2", xlim = c(-600, 600))
text(pca_results$x[,1], pca_results$x[,2],labels=rownames(pca_results$x),cex=0.9,pos=3)
legend("topright",legend=c("DS","WT"),col=c("red","blue"),pch=19)
```
We plot the first two components, labelling each point to the slide array in order to see the level of the batch effect. The plot shows no outliers, the results are clustered by the slide array and a diagonal line of $y=-ax$ divides the samples in the two groups (```WT``` and ```DS```). In the first case we can address the event to a probable presence of the batch effect. One thing that could have been done to avoid this is to randomize the sample in the array.

# **Identification of differently methylated probes**
The following step will be performed using a ```t-test``` to identify differently mathylated probes between ```DS
``` and ```WT``` samples. We will use the matrix with normalized $\beta$ values generated in **Chapter 7**.

We define a function ```t_test_func()``` that allows to perform a t-test to each row of the matrix. This function implements the ```t.test()``` function and then it returns the resulted p-value.
```{r step 9 t-test, message = FALSE, warning = FALSE}
t_test_func <- function(x) {
  t_test <- t.test(x~ csvtargets$Group)
	return(t_test$p.value)
}
pValues_ttest <- apply(preprocessQuantileBeta,1, t_test_func)
```
# **Multiple test correction**
Multiple comparisons (such as the one just performed) has a downside: it allows an increase of the $\alpha$ level, therefore the probability of falsely rejecting the null hypotheses increases while testing. In order to overcome this problem we need to _adjust_ the raw p-values just computed. We have two main correction method we can apply: *Bonferroni correction* {```bonferroni```} and *False Discovery Rate* {```BH```}. We apply both with ```p.adjust()``` function, once with ```bonferroni``` correction method and once with ```BH``` correction method.
```{r step 10 multiple test correction, message = FALSE, warning = FALSE}
NormBeta_pValues <- data.frame(preprocessQuantileBeta, pValues_ttest)
NormBeta_pValues <- NormBeta_pValues[order(NormBeta_pValues$pValues_ttest),]

pValues_raw <- NormBeta_pValues$pValues_ttest
# bonferroni correction applied on raw p-values
pValues_ttest_bonf <- p.adjust(pValues_raw,"bonferroni")
# Benjamini-Hochberg correction applied on raw p-values
pValues_ttest_BH <- p.adjust(pValues_raw, "BH")

NormBeta_pValues <- data.frame(NormBeta_pValues, pValues_ttest_BH, pValues_ttest_bonf)
```
## _significant p-values_

>**Q1**: How many probes do you identify as differentially methylated considering nominal pValues?
>**Q2**: How many after ```Bonferroni``` correction?
>**Q3**: How many after ```BH``` correction?

To answer those questions we need to find all those p-values that are smaller than a given threshold. Our threshold is set at ```0.05``` value.
```{r step 10 detection, message = FALSE, warning = FALSE}
# Q1
dim(NormBeta_pValues[NormBeta_pValues$pValues_ttest<=0.05,])
# Q2
dim(NormBeta_pValues[NormBeta_pValues$pValues_ttest_bonf<=0.05,])
# Q3
dim(NormBeta_pValues[NormBeta_pValues$pValues_ttest_BH<=0.05,])
```
### Table recap
|p-Values|Number of significant samples |
|:-------:|:-----------------:|
|Raw| 39186|
|Bonferroni correction | 0|
|BH correction| 0|

As we can see from the recap table above, the significant probes found taking into account raw p-value are ```39186```, while for both _Bonferroni_ and _BH_ corrected p-values there were no significant probe found.

```{r  step 10 boxplot, fig.width=7, fig.height=7, fig.cap = " **Figure 7** : Boxplot of t-test results.", message = FALSE, warning = FALSE}
ttest_df <- data.frame(pValues_raw, pValues_ttest_BH, pValues_ttest_bonf)
boxplot(ttest_df, ylim = c(-0.1, 1.1), col = c("green", "orange", "cyan"))
```

# **Heatmap**

Heatmaps are useful tools to observe methylation levels in samples, different colors express different level of
 methylation. Similar expressions will cluster together and will have similar color. To generate the heatmap we use the
  ```heatmap.2()``` function with ```complete```, ```single``` and ```average``` linkage on the top 100
   differentially methylated probes.

```{r step 11 - complete heatmaps, fig.height = 8, fig.width = 8, fig.cap = "**Figure 8**: heatmap with complete linkage method", message = FALSE, warning = FALSE, results = 'hide'}
library(gplots)
library(RColorBrewer)
library(gridExtra)

# taking only beta values of top 100 differentialy methylated probes
heatmap = as.matrix(NormBeta_pValues[1:100, 1:8])
colorbar <- c("blue","blue","red","red","red","blue", "blue", "red")

palette <- heat.colors(100)

# complete linkage
heatmap.2(heatmap, col= palette , Rowv=T, Colv=T, dendrogram="both", key=T, ColSideColors=colorbar, labCol = TRUE, density.info="none", trace="none", scale="none", symm=F, main = "Norm beta values (Method: complete linkage)")
text(x = c(0.4, 0.78), y = 0.77,
     labels = c('DS', 'WT'),
     col = 'black',
     cex = 0.8,
     las = 2,
     xpd = TRUE)
```
```{r step 11 - single heatmaps, fig.height = 8, fig.width = 8, fig.cap = "**Figure 9**: heatmap with single linkage method", message = FALSE, warning = FALSE, results = 'hide'}
# single linkage
heatmap.2(heatmap, col= palette , Rowv=T, Colv=T, hclustfun = function(x) hclust(x, method = 'single'), dendrogram ="both", key=T, ColSideColors=colorbar, labCol = TRUE, density.info="none", trace="none", scale="none", symm=F, main = "Norm Beta values (Method: single linkage)")
text(x = c(0.4, 0.78), y = 0.774,
     labels = c('DS', 'WT'),
     col = 'black',
     cex = 0.8,
     las = 2,
     xpd = TRUE)
```
```{r step 11 - average heatmaps, fig.height = 8, fig.width = 8, fig.cap = "**Figure 10**: heatmap with average linkage method", message = FALSE, warning = FALSE, results = 'hide'}
# averge linkage
heatmap.2(heatmap, col= palette, Rowv=T, Colv=T, hclustfun = function(x) hclust(x,method = 'average'), dendrogram ="both", key=T, ColSideColors=colorbar, labCol = TRUE, density.info="none", trace="none", scale="none", symm=F, main = "Norm Beta values (Method: average linkage)")
text(x = c(0.4, 0.78), y = 0.774,
     labels = c('DS', 'WT'),
     col = 'black',
     cex = 0.8,
     las = 2,
     xpd = TRUE)

```
As we can see in **Figures 8, 9** and **10** different linkage methods don't influence the results that much: both groups maintain the clusterization, there are no visible difference in methylation level besides some cases where some probe resulted more methylated than others. The latter case is most prominent in the case of ```cg12012426``` that results highly methylated for ```DS``` but very low methylation in the other group.

# **Additional plots**

Additional plots such as **Volcano plot** and **Manhattan plot** are useful to better visualize results of differential methylation analysis.

## _Volcano plot_
First we define the datasets: we extract normalized $\beta$ values produced in the normalization step and stored in
 ```NormBeta_pValues```, then we compute ```delta``` from the mean  of ```DS``` and ```WT``` groups and plot the
  **Volcano plot**. In order to compare the results we decided to plot three volcano plots: one with    ```raw_pValue```, one with ```bonf_pValue``` and one with ```BH_pValue```. We highlight those probes with
    nominal p-value smaller than ```0.05``` threshold and delta > ```0.01```.
```{r step 12 - Volcano plot, fig.width=10, fig.height=4, fig.cap = "**Figure 11**: (from left) Volcano plot with raw p-value, bonferroni corrected p-value and BH corrected p-value", message = FALSE, warning = FALSE,results='hide'}
NormBeta <- NormBeta_pValues[, 1:8]

# DS
NormBeta_DS <- NormBeta[, csvtargets$Group == "DS"]
meanNormBeta_DS <- apply(NormBeta_DS, 1, mean)
# WT
NormBeta_WT <- NormBeta[, csvtargets$Group == "WT"]
meanNormBeta_WT <- apply(NormBeta_WT, 1, mean)
# delta
delta <- meanNormBeta_DS-meanNormBeta_WT

# Volcano plot raw data
VolcPlot_raw <- data.frame(delta, -log10(NormBeta_pValues$pValues_ttest))
Highlight_raw <- VolcPlot_raw[abs(VolcPlot_raw[, 1])>0.1 & VolcPlot_raw[, 2]>(-log10(0.05)),]
# Volcano plot bonf corrected data
VolcPlot_bonf <- data.frame(delta, -log10(NormBeta_pValues$pValues_ttest_bonf))
Highlight_bonf <- VolcPlot_bonf[abs(VolcPlot_bonf[, 1])>0.1 & VolcPlot_bonf[, 2]>(-log10(0.05)),]
# Volcano plot BH corrected data
VolcPlot_BH <- data.frame(delta, -log10(NormBeta_pValues$pValues_ttest_BH))
Highlight_BH <- VolcPlot_BH[abs(VolcPlot_BH[,1])>0.1 & VolcPlot_BH[,2]>(-log10(0.05)),]

par(mfrow = c(1,3))
# plot Volcano plot raw
plot(VolcPlot_raw[,1], VolcPlot_raw[,2], pch=16, cex=0.5, ylab="raw p-Values (-log)", xlab="delta",ylim = c(0, 7)) -log10(0.05)
abline(a=-log10(0.05),b=0,col="red")
points(Highlight_raw[,1], Highlight_raw[,2],pch=16,cex=0.7,col="red")

# plot Volcano plot bonf
plot(VolcPlot_bonf[,1], VolcPlot_bonf[,2], pch=16,cex=0.5, ylab="bonf corrected p-Values (-log)", xlab="delta", ylim = c(0, 7)) -log10(0.05)
abline(a=-log10(0.05),b=0,col="red")
points(Highlight_bonf[,1], Highlight_bonf[,2],pch=16,cex=0.7,col="red")

# plot Volcano plot BH
plot(VolcPlot_BH[,1], VolcPlot_BH[,2], pch=16,cex=0.5, ylab="BH corrected p-Values (-log)", xlab="delta", ylim = c(0, 7)) -log10(0.05)
abline(a=-log10(0.05),b=0,col="red")
points(Highlight_BH[,1], Highlight_BH[,2],pch=16,cex=0.7,col="red")
```

## _Manhattan plot_
```{r step 12 - Manhattan plot, fig.width=10, fig.height=4, fig.cap = "**Figure 12**: (from left) Manhattan plot with raw p-value, bonferroni corrected p-value and BH corrected p-value", message = FALSE, warning = FALSE, results = 'hide'}
library(gap)
NormBeta_pValues_inv <-data.frame(rownames(NormBeta_pValues), NormBeta_pValues)
colnames(NormBeta_pValues_inv)[1] <- "IlmnID"

# we need to annotate the data
NormBeta_pValues_ann <- merge(NormBeta_pValues_inv, Illumina450Manifest_clean, by="IlmnID")
NormBeta_pValues_ann <- droplevels(NormBeta_pValues_ann)

Manhattan_raw <- data.frame(NormBeta_pValues_ann$CHR, NormBeta_pValues_ann$MAPINFO, NormBeta_pValues_ann$pValues_ttest)
Manhattan_bonf <- data.frame(NormBeta_pValues_ann$CHR, NormBeta_pValues_ann$MAPINFO, NormBeta_pValues_ann$pValues_ttest_bonf)
Manhattan_BH <- data.frame(NormBeta_pValues_ann$CHR, NormBeta_pValues_ann$MAPINFO, NormBeta_pValues_ann$pValues_ttest_BH)

Manhattan_raw$NormBeta_pValues_ann.CHR <- factor(Manhattan_raw$NormBeta_pValues_ann.CHR, levels = c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","X","Y"))
Manhattan_bonf$NormBeta_pValues_ann.CHR <- factor(Manhattan_bonf$NormBeta_pValues_ann.CHR, levels = c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","X","Y"))
Manhattan_BH$NormBeta_pValues_ann.CHR <- factor(Manhattan_BH$NormBeta_pValues_ann.CHR, levels = c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","X","Y"))


palette <- rainbow(24)

par(mfrow = c(1, 3))
# Manhattan plot raw p-values
mhtplot(Manhattan_raw,control=mht.control(colors=palette))
axis(2,cex=0.5)
abline(a=-log10(0.05),b=0)
# Manhattan plot bonf corrected p-values
mhtplot(Manhattan_bonf,control=mht.control(colors=palette), ylim = c(0,7))
axis(2,cex=0.5)
abline(a=-log10(0.05),b=0)
# Manhattan plot BH corrected p-values
mhtplot(Manhattan_BH,control=mht.control(colors=palette), ylim = c(0,7))
axis(2,cex=0.5)
abline(a=-log10(0.05),b=0)
```
Both sets of plots (**Figures 11 and 12**) are consistent with previous analysis, indeed we can highlight probes below
 significant threshold only in non corrected p-values case, while no probes are highlighted in both bonferroni and BH
  corrected p-value.


# **Focus on chromosome 21**
Down syndrome is caused by the trisomy of chromosome 21, in this final step we will perform an in depht analysis to
 visualize if there are any difference in methylation between ```DS``` sample group and ```WT``` specifically on
  chromosome 21. We will perform this task by plotting normalized $\beta$ values by the annotated object we generated
   in the previous step and cross-reference it with the normalized $\beta$ values in order to retain only those
    associated to chromose 21. Then we subset the beta values in ```WT``` and ```DS``` as we did in the **chapter 6
    **, we compute ```mean``` and ```densities``` for both subsets and we plot it (**Figure 13**).
```{r - optional analysis chr21, fig.cap = "**Figure 13**: Chromosome 21 beta density distribution", message = FALSE, warning = FALSE, results = "hide"}
chr21 <- NormBeta_pValues_ann[NormBeta_pValues_ann$CHR == "21",]
chr21 <- droplevels(chr21)
beta_21 <- preprocessQuantileBeta[rownames(preprocessQuantileBeta) %in% chr21$IlmnID,]

#WT
beta_wt_21 <- beta_21[, colnames(beta_21) %in% WT]
mean_beta_wt_21 <- apply(beta_wt_21, 1, mean)
d_mean_beta_wt_21 <- density(mean_beta_wt_21, na.rm = T)

#DS
beta_ds_21 <- beta_21[, colnames(beta_21) %in% DS]
mean_beta_ds_21 <- apply(beta_ds_21, 1, mean)
d_mean_beta_ds_21 <- density(mean_beta_ds_21, na.rm = T)

plot(d_mean_beta_wt_21, col = "green", main = "Normalized Beta value distribution CHR21")
lines(d_mean_beta_ds_21, col = "red")
legend("topright", legend=c("WT", "DS"),
       col=c("green", "red"), lty=1:1, cex=0.5)
```
We can note that there isn't much difference between the two groups, the distributions follow almost the same curve
 trend besides ```WT``` distribution having a slightly higher peak than the ```DS``` curve on high values for $\beta$.

>**Q**: How many differentially methylated probes do you find on chromosome 21?

To answer this question we will simply use the ```dim()``` function on the dataset that contains p-values
 associated to chromosome 21. We use ```raw_pvalues```, and both corrected versions ```bonf_pValues``` and
  ```BH_pValues```.
```{r optional - differentially methylated chr21}
dim(chr21[chr21$pValues_ttest <= 0.05,])
dim(chr21[chr21$pValues_ttest_BH <= 0.05,])
dim(chr21[chr21$pValues_ttest_bonf<= 0.05,])
```
## _Table recap_

Results are recapped in the following table:

| p-values | differentially methylated probes|
|:--------:|:--------------------------------|
|raw | 598 |
|bonferroni | 0|
|BH | 0|

Consistently with the previous results, we found differentially methylated probes only if taking into account raw p
-values.